File: javascript/test/pedido.test.js
const { expect } = require('chai');
const { Pedido } = require('../src/pedido');

describe('Pedido con aserciones fluidas', () => {
  it('conserva el orden de los productos', () => {
    // Arrange
    const pedido = new Pedido(
      'Ana',
      ['Libro', 'Cuaderno', 'Lápiz'],
      100,
      10,
    );

    // Act
    const productos = pedido.productos;

    // Assert: colección fluida
    expect(productos, 'productos del pedido de Ana')
      .to.deep.equal(['Libro', 'Cuaderno', 'Lápiz'])
      .and.not.include('Borrador');
  });

  it('expone sus propiedades', () => {
    // Arrange
    const pedido = new Pedido('Luis', ['Teclado'], 200, 5);

    // Act & Assert: propiedades del objeto
    expect(pedido).to.include({
      cliente: 'Luis',
      subtotal: 200,
      descuento: 5,
    });
  });

  it('aplica el descuento al total', () => {
    // Arrange
    const pedido = new Pedido('Sara', ['Monitor'], 100, 10);

    // Act
    const total = pedido.calcularTotal();

    // Assert: comparación numérica con tolerancia
    expect(total, 'total después del descuento').to.be.closeTo(90, 0.001);
  });

  it('genera un resumen legible', () => {
    // Arrange
    const pedido = new Pedido('Mario', ['Mouse', 'Base'], 80);

    // Act
    const resumen = pedido.generarResumen();

    // Assert: texto fluido
    expect(resumen)
      .to.match(/^Pedido de Mario/)
      .and.include('2 producto(s)')
      .and.not.include('undefined');
  });

  it('explica el error cuando el descuento es inválido', () => {
    // Act & Assert: tipo y mensaje de la excepción
    expect(() => new Pedido('Laura', ['Agenda'], 50, 110))
      .to.throw(Error, 'El descuento debe estar entre 0 y 100');
  });
});

File: java/src/test/java/com/ejemplo/PedidoTest.java
package com.ejemplo;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.Assertions.within;

import java.util.List;
import org.junit.jupiter.api.Test;

class PedidoTest {

    @Test
    void productos_deberianConservarElOrden() {
        // Arrange
        Pedido pedido = new Pedido(
            "Ana",
            List.of("Libro", "Cuaderno", "Lápiz"),
            100.0,
            10.0
        );

        // Act
        List<String> productos = pedido.getProductos();

        // Assert: colección fluida
        assertThat(productos)
            .as("productos del pedido de Ana")
            .containsExactly("Libro", "Cuaderno", "Lápiz")
            .doesNotContain("Borrador");
    }

    @Test
    void pedido_deberiaExponerSusPropiedades() {
        // Arrange
        Pedido pedido = new Pedido("Luis", List.of("Teclado"), 200.0, 5.0);

        // Act & Assert: propiedades del objeto
        assertThat(pedido)
            .extracting("cliente", "subtotal", "descuento")
            .containsExactly("Luis", 200.0, 5.0);
    }

    @Test
    void total_deberiaAplicarElDescuento() {
        // Arrange
        Pedido pedido = new Pedido("Sara", List.of("Monitor"), 100.0, 10.0);

        // Act
        double total = pedido.calcularTotal();

        // Assert: comparación numérica con tolerancia
        assertThat(total)
            .as("total después del descuento")
            .isCloseTo(90.0, within(0.001));
    }

    @Test
    void resumen_deberiaSerLegible() {
        // Arrange
        Pedido pedido = new Pedido("Mario", List.of("Mouse", "Base"), 80.0, 0.0);

        // Act
        String resumen = pedido.generarResumen();

        // Assert: texto fluido
        assertThat(resumen)
            .startsWith("Pedido de Mario")
            .contains("2 producto(s)")
            .doesNotContain("null");
    }

    @Test
    void descuentoInvalido_deberiaExplicarElProblema() {
        // Act & Assert: excepción, tipo y mensaje
        assertThatThrownBy(
            () -> new Pedido("Laura", List.of("Agenda"), 50.0, 110.0)
        )
            .isInstanceOf(IllegalArgumentException.class)
            .hasMessage("El descuento debe estar entre 0 y 100");
    }
}
